### PR TITLE
Secure upgrade

### PR DESCRIPTION
**Pull request by ycoheNvidia**
#### What I did
Added support for secure upgrade

#### How I did it
It includes image signing during build (in sonic buildimage repo) and verification during image install (in sonic-utilities).
HLD can be found in the following PR: https://github.com/sonic-net/SONiC/pull/1024
#### How to verify it
Feature is used to allow image was not modified since built from vendor. During installation, image can be verified with a signature attached to it.
In order for image verification - image must be signed - need to provide signing key and certificate (paths in SECURE_UPGRADE_DEV_SIGNING_KEY and SECURE_UPGRADE_DEV_SIGNING_CERT in rules/config) during build , and during image install, need to enable secure boot flag in bios, and signing_certificate should be available in bios.

#### Feature dependencies
In order for this feature to work smoothly, need to have secure boot feature implemented as well.
The Secure boot feature will be merged in the near future.

sonic-buildimage PR: https://github.com/sonic-net/sonic-buildimage/pull/11862

### CODE REVIEW

**Review by Yarden-Z — DISMISSED** [review]
Looks good

**Comment by liat-grozovik** [comment]
@yxieca could you please help to review or suggest a reviewer?

**Comment by ycoheNvidia** [comment]
@yxieca  @liat-grozovik I reviewed the scan of bendit. There are 70 high severity issues that are not related to my code, but from the original master. 
Regarding my code, there are only few medium issues, and don't think they are required to be resolved. How do you want to proceed?

**Comment by liat-grozovik** [comment]
as of today bandit is not required. it is still under work.
once the pr is approved by @xumia  we will be able to merge.

**Comment by Yarden-Z** [comment]
> as of today bandit is not required. it is still under work. once the pr is approved by @xumia we will be able to merge.

@xumia Is there any update on this PR?

File: scripts/verify_image_sign_test.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
Comment (xumia) [inline scripts/verify_image_sign_test.sh:1]: Can we move all the test script in the tests folder?

Comment (ycoheNvidia) [reply scripts/verify_image_sign_test.sh:1]: Where do you suggest to move these test to?

Comment (xumia) [reply scripts/verify_image_sign_test.sh:1]: The script scripts/verify_image_sign_test.sh is used by tests/sign_and_verify_test.py, can we move the script verify_image_sign_test.sh to the tests/ folder?

File: scripts/verify_image_sign_common.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    RESULT="CMS Verified OK"
+    if  [ -d "${TMP_DIR}" ]; then rm -rf ${TMP_DIR}; fi
+    echo "verification ok:$RESULT"
+    # No need to continue.
+    # Exit without error if any success signature verification.
+    return 0
+fi
+
+if  [ -d "${TMP_DIR}" ]; then rm -rf ${TMP_DIR}; fi
+return 1
```
Comment (xumia) [inline scripts/verify_image_sign_common.sh:33]: How about adding 4 space characters indentation for the scripts inside of the function verify_image_sign_common?

**Review by xumia — APPROVED** [review]

**Comment by Staphylo** [comment]
This change will break `sonic-installer install` for all Aboot and Uboot platforms.
It implements `verify_image_sign` solely for Grub not even declaring it in its base class while still using it in common code.
Any reason why this code is not using the existing `verify_secureboot_image` like it's done here? https://github.com/sonic-net/sonic-utilities/blob/c57c3fadc493d219027c91ff7c26e3d810ef3693/sonic_installer/bootloader/aboot.py#L199
@xumia @qiluo-msft this change needs to be reverted or fixed ASAP before it makes its way into sonic-buildimage.
Could you also add me as a notifiee everytime someone changes `sonic-utilities/sonic_installer/bootloader/` ?

**Comment by xumia** [comment]
@Staphylo , thanks for your comment.
@ycoheNvidia, @liat-grozovik , could you please help fix it ASAP? It has impact on the Aboot and Uboot platforms.

**Comment by ycoheNvidia** [comment]
> @Staphylo , thanks for your comment. @ycoheNvidia, @liat-grozovik , could you please help fix it ASAP? It has impact on the Aboot and Uboot platforms.

Will upload a patch today, thanks for the vigilance

**Comment by ycoheNvidia** [comment]
> @Staphylo , thanks for your comment. @ycoheNvidia, @liat-grozovik , could you please help fix it ASAP? It has impact on the Aboot and Uboot platforms.

added a fix in https://github.com/sonic-net/sonic-utilities/pull/2646

**Comment by StormLiangMS** [comment]
@liat-grozovik this PR caused aboot failure with 202205 and 202012 images, and block pipeline test, will revert first, pls re-submit when this issue get fixed.

**Comment by ycoheNvidia** [comment]
> @liat-grozovik this PR caused aboot failure with 202205 and 202012 images, and block pipeline test, will revert first, pls re-submit when this issue get fixed.

we have a fix in https://github.com/sonic-net/sonic-utilities/pull/2646

**Pull request merged**